### PR TITLE
Replace axios with @actions/http-client

### PR DESCRIPTION
**Pull request by mctofu**
Axios has [issues](https://github.com/axios/axios/issues/4531) with proxies. Rather than trying to workaround these problems this swaps our http client to [`@actions/http-client`](https://github.com/actions/toolkit/tree/main/packages/http-client) which is designed to work in actions environments.

The swap was pretty straightforward except the built in retry behavior was not as robust. We needed it to handle additional 5xx codes so I've implemented retries manually.

### CODE REVIEW

**Review by honeyankit — APPROVED** [review]

File: src/api-client.ts [hunk 1]
```diff
@@ -43,48 +43,65 @@ export class ApiClient {
...
+      )
+      if (res.statusCode !== 200) {
         throw new JobDetailsFetchingError(
-          `fetching job details: unexpected status code: ${res.status}`
+          `fetching job details: unexpected status code: ${
+            res.statusCode
+          }: ${JSON.stringify(res.result)}`
+        )
+      }
+      if (res.result == null) {
```
Comment (abdulapopoola) [inline src/api-client.ts:59]: consider idiomatic JS

```
if(!res.result) {
```

File: src/api-client.ts [hunk 2]
```diff
@@ -43,48 +43,65 @@ export class ApiClient {
...
+        throw new JobDetailsFetchingError(
+          `fetching job details: missing response`
         )
       }
 
-      return res.data.data.attributes
+      return res.result.data.attributes
```
Comment (abdulapopoola) [inline src/api-client.ts:65]: can any of these be `falsy` and throw a null ref?

File: src/api-client.ts [hunk 3]
```diff
@@ -43,48 +43,65 @@ export class ApiClient {
...
         throw new JobDetailsFetchingError(
-          `fetching job details: ${(error as Error).message}`
+          `fetching job details: unexpected status code: ${error.statusCode}: ${error.message}`
+        )
+      } else if (error instanceof Error) {
+        throw new JobDetailsFetchingError(
+          `fetching job details: ${error.name}: ${error.message}`
         )
       }
+      throw error
```
Comment (abdulapopoola) [inline src/api-client.ts:78]: consider adding semicolons at the end to avoid weird JS automatic semicolon injection bugs

File: src/api-client.ts [hunk 4]
```diff
@@ -93,53 +110,84 @@ export class ApiClient {
...
+    }
+  }
+
+  private async getJsonWithRetry<T>(
+    url: string,
+    token: string
+  ): Promise<TypedResponse<T>> {
+    let attempt = 1
+
+    const execute = async (): Promise<TypedResponse<T>> => {
```
Comment (abdulapopoola) [inline src/api-client.ts:164]: how does the actions team handle retries? Do they have some utility we can reuse?

**Comment by abdulapopoola** [comment]
[Aside] will we need to remove `axios` [totally?](https://github.com/github/dependabot-action/blob/2159f6cc2ed7c620ce61d32ff1a1f2509371fcb9/package.json#L37)

File: src/api-client.ts (hunk 2 above)
Comment (mctofu) [reply src/api-client.ts:65]: It wouldn't be expected so a null ref failing the job would be OK.

File: src/api-client.ts (hunk 3 above)
Comment (mctofu) [reply src/api-client.ts:78]: Hm, our formatter does not allow that

File: src/api-client.ts (hunk 4 above)
Comment (mctofu) [reply src/api-client.ts:164]: It's embedded in the library and not exported: https://github.com/actions/toolkit/blob/fe3e7ce9a7f995d29d1fcfd226a32bca407f9dc8/packages/http-client/src/index.ts#L444-L450

**Pull request merged**